Paper: PyCID: A Python Library for Causal Influence Diagrams

Jamesfox1 opened the pull request:
If you are creating this PR in order to submit a draft of your paper,
see http://procbuild.scipy.org/ for logs generated by the build
process.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#build-your-paper)
for more information.


*Editor*: <!--editor-->@deniederhut<!--end-editor-->

*Reviewers*: <!--reviewers-list-->@deniederhut<!--end-reviewers-list-->

Jamesfox1 committed:
adding names

Jamesfox1 committed:
adding some writing

Jamesfox1 committed:
first draft

RyanCarey committed:
slight wording changes

RyanCarey committed:
more typos and minor changes

Jamesfox1 committed:
minor

Jamesfox1 committed:
Merge branch '2021' of https://github.com/jamesfox1/scipy_proceedings into 2021

Jamesfox1 committed:
fix spelling typos

Jamesfox1 committed:
edit

Jamesfox1 committed:
change intro

Jamesfox1 committed:
edits

Jamesfox1 committed:
fix merge

Jamesfox1 committed:
minor

Jamesfox1 committed:
delete folder

Jamesfox1 committed:
minor

Jamesfox1 committed:
edits

scoobies commented:
Hi!, I'm @scoobies, a friendly bot.

Type ```@scoobies help``` to discover how I can help you.

Jamesfox1 commented:
@scoobies help

scoobies commented:
Hello @Jamesfox1, here are the things you can ask me to do:

```

# List all available commands
@scoobies help

# Show our community Code of Conduct and Guidelines
@scoobies code of conduct

# Assign a user as the reviewer N of this submission (where N=1,2...)
@scoobies assign @username as reviewer 2

# Remove the user assigned as reviewer N of this submission (where N=1,2...)
@scoobies remove reviewer 2

# Add to this issue's reviewers list
@scoobies add @username to reviewers

# Remove from this issue's reviewers list
@scoobies remove @username from reviewers

# Assign a user as the editor of this submission
@scoobies assign @username as editor

# Remove the editor assigned to this submission
@scoobies remove editor

# Send an invitation to a user to collaborate in the review
@scoobies invite @username

# Add a user to this issue's assignees list
@scoobies add assignee: @username

# Remove a user from this issue's assignees list
@scoobies remove assignee: @username

# Builds paper
@scoobies build paper

# Checks build status
@scoobies build status

# Check the references of the paper for missing DOIs
# Optionally, it can be run on a non-default branch 
@scoobies check references
@scoobies check references from custom-branch-name

# Label issue with: paper
@scoobies mark paper

# Label issue with: needs-review. Remove labels: unready, needs-more-review, pending-comment, ready
@scoobies mark needs review

# Label issue with: needs-more-review. Remove labels: unready, needs-review, pending-comment, ready
@scoobies mark needs more review

# Label issue with: pending-comment. Remove labels: unready, needs-review, needs-more-review, ready
@scoobies mark pending comment

# Label issue with: ready. Remove labels: unready, needs-review, needs-more-review, pending-comment
@scoobies mark ready

# Label issue with: unready. Remove labels: needs-review, needs-more-review, pending-comment, ready
@scoobies mark not ready

# Label issue with: does_not_build:server
@scoobies mark server fail

# Remove labels: does_not_build:server
@scoobies mark server success

```

deniederhut commented:
@scoobies assign @deniederhut as editor

scoobies commented:
Assigned! @deniederhut is now the editor

Jamesfox1 committed:
Tom's edits

deniederhut commented:
@scoobies add @deniederhut to reviewers

scoobies commented:
@deniederhut added to the reviewers list!

deniederhut commented:
Hey @Jamesfox1 ! We've had a hard time finding reviewers for this paper, so I'll be stepping in to provide feedback. I should have some comments for you this weekend.

deniederhut commented on papers/.vscode/settings.json:
```diff
@@ -0,0 +1,3 @@
+{
```
Please don't modify the repo outside of your paper's directory

deniederhut commented on papers/james_fox/.vscode/settings.json:
```diff
@@ -0,0 +1,4 @@
+{
```
Did you mean to commit this?

Jamesfox1 replied:
No, sorry!

deniederhut commented on papers/james_fox/mybib.bib:
```diff
@@ -0,0 +1,263 @@
+  year={2021}
+}
+
+@inproceedings{ankan2015pgmpy,
```
You can find the citation for this paper here: https://conference.scipy.org/proceedings/scipy2015/bib/ankur_ankan.bib

Jamesfox1 replied:
Thanks!

deniederhut commented on papers/james_fox/mybib.bib:
```diff
@@ -0,0 +1,263 @@
+  year={2020}
+}
+
+@article{everitt2019reward,
```
We would prefer the journal citation for this paper

deniederhut commented on papers/james_fox/mybib.bib:
```diff
@@ -0,0 +1,263 @@
+  year={2019}
+}
+
+@inproceedings{holtman2020towards,
```
This is missing a DOI

deniederhut commented on papers/james_fox/mybib.bib:
```diff
@@ -0,0 +1,263 @@
+  year={2021}
+}
+
+@article{kusner2017counterfactual,
```
We would prefer the conference citation for this paper

deniederhut commented on papers/james_fox/mybib.bib:
```diff
@@ -0,0 +1,263 @@
+}
+
+
+@article{howard2005influence,
```
This is missing a DOI

deniederhut commented on papers/james_fox/mybib.bib:
```diff
@@ -0,0 +1,263 @@
+}
+
+
+@article{borgonovo2016sensitivity,
```
This is missing a DOI

deniederhut commented on papers/james_fox/mybib.bib:
```diff
@@ -0,0 +1,263 @@
+  year={2007}
+}
+
+@techreport{hagberg2008exploring,
```
Was this a SciPy paper? http://conference.scipy.org/proceedings/scipy2008/paper_2/

deniederhut commented on papers/james_fox/mybib.bib:
```diff
@@ -0,0 +1,263 @@
+  year={2020}
+}
+
+@article{hunter2007matplotlib,
```
This is missing a DOI

deniederhut commented on papers/james_fox/mybib.bib:
```diff
@@ -0,0 +1,263 @@
+  publisher={IEEE Computer Society}
+}
+
+@Article{harris2020array,
```
This is missing a DOI

deniederhut commented on papers/james_fox/mybib.bib:
```diff
@@ -0,0 +1,263 @@
+          Warren Weckesser and Hameer Abbasi and Christoph Gohlke and
+          Travis E. Oliphant},
+ year={2020},
+ month= sep,
```
I don't think we need the month, but presumably this should be `{September}`?

deniederhut commented on papers/james_fox/pycid.rst:
```diff
@@ -0,0 +1,666 @@
+:author: James Fox
+:email: james.fox@cs.ox.ac.uk
+:institution: University of Oxford
```
I think we're missing a `:corresponding:` tag here

deniederhut commented on papers/james_fox/pycid.rst:
These lines are a couple of characters over the column width. We might need to do something like

```python
U1=lambda d1, d2: agent1_payoff[
    d1_domain.index(d1), d2_domain.index(d2)
],
```

deniederhut commented on papers/james_fox/pycid.rst:
This is roughly where the 8 page mark is in the rendered pdf. Are they are a couple of sentences we could remove / shorten somewhere to stay within the limit?

deniederhut commented on papers/james_fox/pycid.rst:
```diff
@@ -0,0 +1,666 @@
+Introduction
+------------
+   
+Influence-diagrams (IDs) are used to represent and analyse decision making situations under uncertainty (:cite:`howard2005influence`, :cite:`miller1976development`). Like Bayesian Networks, IDs have at their core a directed acyclic graph, but IDs also specify decision and utility nodes. Relationships between variables are specified with conditional probability distributions. When these are specif ...
```
You can expect your audience at SciPy to be generally scientifically literate, but possibly unaware of things like causal inference or path analysis. Domain specific terms should be defined when they are used. 

You may also find it helpful to include a concrete example to ground your readers' intuition about what is being represented. Can you make a causal influence model that describes whether or not I buy an ice cream at the corner store? What is my utility node in that context? What is my decision rule? Which nodes are random or conditioned on other nodes?

Jamesfox1 replied:
I have now incorporated a simple example into the introduction, which is then threaded through the first few sections. I have also (as far aas I can tell) now defined all new terms as and when they are used

deniederhut commented on papers/james_fox/pycid.rst:
```diff
@@ -0,0 +1,666 @@
+   
+Influence-diagrams (IDs) are used to represent and analyse decision making situations under uncertainty (:cite:`howard2005influence`, :cite:`miller1976development`). Like Bayesian Networks, IDs have at their core a directed acyclic graph, but IDs also specify decision and utility nodes. Relationships between variables are specified with conditional probability distributions. When these are specif ...
+   
+Statistical and causal IDs have already shown promise for a wide variety of applications from the design of AI systems with safe and fair incentives (:cite:`everitt2021agent`, :cite:`carey2020incentives`, :cite:`everitt2019reward`, :cite:`holtman2020towards`, :cite:`everitt2019modeling`, :cite:`langlois2021rl`, and :cite:`cohen2020asymptotically`) to applications in business and medical decision  ...
```
Same comment as above - what does it mean to "show promise for ... medical decision-making"? What does causal inference give us that a doctor doesn't? Why would someone prefer a hospital that uses this tool? Why would someone use this tool for their own decision-making?

Could we include links to reference sites for pyAgrum and PySMILE? I've found three different python packages named pysmile on github, and none of them appear to be the one you mean.

Jamesfox1 replied:
I have hopefully now addressed these questions

deniederhut commented on papers/james_fox/pycid.rst:
```diff
@@ -0,0 +1,666 @@
+   
+Statistical and causal IDs have already shown promise for a wide variety of applications from the design of AI systems with safe and fair incentives (:cite:`everitt2021agent`, :cite:`carey2020incentives`, :cite:`everitt2019reward`, :cite:`holtman2020towards`, :cite:`everitt2019modeling`, :cite:`langlois2021rl`, and :cite:`cohen2020asymptotically`) to applications in business and medical decision  ...
+ 
+Consequently, this paper introduces *PyCID* [#]_, a Python library built upon *pgmpy* and *NetworkX* :cite:`hagberg2008exploring`, that concentrates on this need for an implementation of IDs and IMs (including their causal and multi-agent variants) and provides researchers and practitioners with convenient methods for analysing decision-making situations. *PyCID* can solve single-agent IMs, find  ...
```
Could we include definitions or examples for interventions and incentives? I think everyone will have a folk understanding of what these mean, but you are using them here as technical terms.

Jamesfox1 replied:
done (and provided examples)

deniederhut commented on papers/james_fox/pycid.rst:
```diff
@@ -0,0 +1,666 @@
+ 
+Consequently, this paper introduces *PyCID* [#]_, a Python library built upon *pgmpy* and *NetworkX* :cite:`hagberg2008exploring`, that concentrates on this need for an implementation of IDs and IMs (including their causal and multi-agent variants) and provides researchers and practitioners with convenient methods for analysing decision-making situations. *PyCID* can solve single-agent IMs, find  ...
+ 
+The next section will describe *PyCID*'s architecture along with installation instructions and some necessary background knowledge. We then move to showcasing some of *PyCID*'s features through applications to incentives and analysing games. In the *Causal Influence Diagrams* section, we demonstrate how to instantiate a simple CID in *PyCID* before demonstrating how to find the nodes which admit  ...
```
Did I miss what the MA stands for in MACIDs?

Jamesfox1 replied:
Thanks! - It means "multi-agent"! I have edited this

deniederhut commented on papers/james_fox/pycid.rst:
```diff
@@ -0,0 +1,666 @@
+Inherited Classes
++++++++++++++++++
+   
+.. figure:: pycid.PNG
```
I don't know how helpful a graphic of the file layout is. This might be a place where we can trim the content down a bit to stay within the page limits?

Jamesfox1 replied:
It's currently still included, say if you think it should still be removed

deniederhut commented on papers/james_fox/pycid.rst:
```diff
@@ -0,0 +1,666 @@
+The */Core* folder
+++++++++++++++++++
+
+*PyCID*'s base class is ``CausalBayesianNetwork``. This class inherits from *pgmpy*'s ``BayesianModel`` and represents a *causal Bayesian network*. Concretely, it extends ``BayesianModel`` by adding the ability to query the effect of *causal interventions*. It also adds methods for determining the expected value of a variable for a given *context* (again under an optional *causal intervention*) a ...
```
This is a minor quibble, but abstract/concrete are technical terms in object-oriented programming. Maybe we could say "specifically", or "in particular"?

deniederhut commented on papers/james_fox/pycid.rst:
```diff
@@ -0,0 +1,666 @@
+
+.. [#] *PyCID* is under continued development, so more features will be added over time.
+   
+The */Core* folder
```
Typically, I see the organization of Python packages described in terms of modules and import paths, not folders. E.g. something like "The pycid.core module contains classes for instantiating causal bayesian networks and assigning probability distributions to their nodes... The pycid.analyze model contains functions for determining the effect of causal interventions on previously defined networks... etc".

Jamesfox1 replied:
I've made this change

deniederhut commented on papers/james_fox/pycid.rst:
I don't know how helpful it is to describe the testing practices here, but if you do, please cite pep8 (https://www.python.org/dev/peps/pep-0008/), pep 484 (https://www.python.org/dev/peps/pep-0484/), and pytest (https://github.com/pytest-dev/pytest/blob/main/CITATION)

deniederhut commented on papers/james_fox/pycid.rst:
This is super nitpicky, but technically that is an initializer, not a constructor.

deniederhut commented on papers/james_fox/pycid.rst:
Also super nitpicky, but please include an `import pycid` in this first code example

deniederhut commented on papers/james_fox/pycid.rst:
I'm not sure we need this backslash here

deniederhut commented on papers/james_fox/pycid.rst:
> With RI, we showed that implementing CIDs in *PyCID* can aid the design of fairer systems. With ICI, we demonstrate how *PyCID* can be used to help design safer AI systems.

Given how closely linked these sentences are, would it make sense to have a stronger conjunction than proximity? Like, a semicolon or something?

deniederhut commented on papers/james_fox/pycid.rst:
Probably we want either "of high quality" or "of high, and not poor".

deniederhut commented on papers/james_fox/pycid.rst:
Do we want to have a link to the causal incentives working group here?

deniederhut commented on papers/james_fox/pycid.rst:
What solver is PyCID using?

Jamesfox1 replied:
We wrote it - have clarified that it's a CID class method

deniederhut reviewed (changes requested):
I really enjoyed reading this paper! It sounds like this library is really filling a missing need, and your examples late in the paper were both illustrative and timely.

I've left a few inline comments which mostly concern formatting, technical terminology, and references. The one non-trivial area where I see room for growth is in the introduction / background. A lot of conference attendees may not be familiar with causal inference (this could be their first exposure!), and I'm worried that they'll see the introduction and assume this paper isn't for them. It might help to spend some more time right at the beginning to motivate why this practice matters, and to spend some time giving readers good intuitions about what this tool does and how.

deniederhut commented:
@scoobies mark needs more review

Jamesfox1 committed:
Mike's edits

Jamesfox1 committed:
Mike's edits

Jamesfox1 commented:
> I really enjoyed reading this paper! It sounds like this library is really filling a missing need, and your examples late in the paper were both illustrative and timely.
> 
> I've left a few inline comments which mostly concern formatting, technical terminology, and references. The one non-trivial area where I see room for growth is in the introduction / background. A lot of conference attendees may not be familiar with causal inference (this could be their first exposure!), and I'm worried that they'll see the introduction and assume this paper isn't for them. It might help to spend some more time right at the beginning to motivate why this practice matters, and to spend some time giving readers good intuitions about what this tool does and how.

Many thanks for your detailed comments and suggestions Dillon. I'm going to work through them now. For your more general comment, you're right. I'm now going to include a basic example of an Influence Diagram in the introduction!

Jamesfox1 committed:
update .bib file

Jamesfox1 committed:
delte .vscode folder

deniederhut commented:
@Jamesfox1 to give everyone more time to respond to their reviews, we've pushed back the deadline for authors' final revisions to July 1st. How are the revisions coming along?

Jamesfox1 committed:
edits

Jamesfox1 commented:
@scoobies build paper

scoobies commented:
building paper...

scoobies commented:
{"data"=>{"info"=>"Build for paper Jamesfox1-2021 scheduled.  Note that builds are only executed if the current build attempt is more than 5 minutes old."}, "status"=>"success"}

Jamesfox1 commented:
> @Jamesfox1 to give everyone more time to respond to their reviews, we've pushed back the deadline for authors' final revisions to July 1st. How are the revisions coming along?

Thanks! I've pushed the revisions - hopefully I've addressed all of your suggestions? The main changes are:
1) adding example into the introduction
2) threading that example into the first few sections.
3) adding the actual definitions of VoI and VoC
4) structure of the library is explained in terms of modules and import paths now rather than files and folders 
5) tried to define any concept as soon as soon as it is introduced.

Jamesfox1 committed:
edit cpd description

deniederhut commented:
Awesome! I'll take another look this week-end

deniederhut commented on papers/james_fox/pycid.rst:
I think we either want "both players defecting" or "both players' defections" here.

deniederhut reviewed (changes requested):
Thanks for making all those changes @Jamesfox1 ! I think I've found one more typo, but otherwise we should be good to go

deniederhut commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1017/CBO9780511808432 is OK
- 10.25080/Majora-7b98e3ed-001 is OK
- 10.1007/978-3-030-52152-3_21 is OK
- 10.1287/deca.1050.0020 is OK
- 10.1016/j.ejor.2015.06.032 is OK
- 10.1109/MCSE.2007.55 is OK
- 10.1038/s41586-020-2649-2 is OK

MISSING DOIs

- 10.1007/978-3-540-39879-0_9 may be a valid DOI for title: Real-world applications of influence diagrams
- 10.1016/s0899-8256(02)00544-4 may be a valid DOI for title: Multi-agent influence diagrams for representing and solving games
- 10.1007/s11229-021-03141-4 may be a valid DOI for title: Reward tampering problems and solutions in reinforcement learning: A causal influence diagram perspective
- 10.21236/ada026379 may be a valid DOI for title: Development of automated aids for decision analysis
- 10.2307/j.ctv173f1fh.6 may be a valid DOI for title: Equilibrium points in n-person games
- 10.1609/aaai.v34i03.5628 may be a valid DOI for title: Asymptotically unambitious artificial general intelligence

INVALID DOIs

- None
```

deniederhut commented:
👀  scoobies found a lot more DOIs than I did -- would mind checking if those are correct, and adding them to the bibfile if appropriate?

deniederhut commented:
@scoobies mark pending comment

Jamesfox1 committed:
update dois

Jamesfox1 committed:
delete comments

Jamesfox1 committed:
minor edits

Jamesfox1 commented:
@scoobies build paper

scoobies commented:
building paper...

scoobies commented:
{"data"=>{"info"=>"Build for paper Jamesfox1-2021 scheduled.  Note that builds are only executed if the current build attempt is more than 5 minutes old."}, "status"=>"success"}

Jamesfox1 commented:
@scoobies check references

Jamesfox1 commented:
> 👀 scoobies found a lot more DOIs than I did -- would mind checking if those are correct, and adding them to the bibfile if appropriate?

Thanks - added!

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1017/CBO9780511808432 is OK
- 10.1007/978-3-540-39879-0_9 is OK
- 10.25080/Majora-7b98e3ed-001 is OK
- 10.1007/s11229-021-03141-4 is OK
- 10.1007/978-3-030-52152-3_21 is OK
- 10.1287/deca.1050.0020 is OK
- 10.1016/j.ejor.2015.06.032 is OK
- 10.21236/ada026379 is OK
- 10.2307/j.ctv173f1fh.6 is OK
- 10.1609/aaai.v34i03.5628 is OK
- 10.1109/MCSE.2007.55 is OK
- 10.1038/s41586-020-2649-2 is OK

MISSING DOIs

- None

INVALID DOIs

- 10.1016/s0899-82560200544-4 is INVALID
```

deniederhut commented:
> 10.1016/s0899-82560200544-4 is INVALID

Hm not sure what's going on with scoobies there, but that one resolves on doi.org just fine.

deniederhut reviewed (approved):
Thanks for making all those changes! I think we're good to go

deniederhut commented:
@scoobies mark ready
